### DISCUSSION TITLE
An extension published by Microsoft can use proposed VSCode APIs, but my extension can't?

### DISCUSSION THREAD

**Discussion by tomqwpl** [microsoft/vscode-discussions#899] [Extension Development QnA] [answered]
I'm writing a VSCode extension and wanting to add some "Notebook" support. Using the Microsoft Jupyter extension and then studying how it works, it seems to make use of proposed APIs:
```
    "enabledApiProposals": [
        "notebookDeprecated",
        "notebookMessaging",
        "notebookMime",
        "notebookCellExecutionState",
        "portsAttributes",
        "quickPickSortByLabel",
        "notebookKernelSource",
        "interactiveWindow",
        "notebookControllerAffinityHidden",
        "quickPickItemTooltip",
        "notebookExecution"
    ],

```

Specifically the one I'm interested in here is the "notebookKernelSource" proposal. I want to hook into the process by which users can select a notebook kernel in the same way that the Jupyter extension does.

According to [this page](https://code.visualstudio.com/api/advanced-topics/using-proposed-api) in order to use proposed APIs you need to use the insiders edition and then either start in development mode or add the `--enable-proposed-api` command line option. 
I've just installed the MS Jupyter extension normally, no insiders version of VSCode, and I'm running VSCode normally.

So conclusion is that Microsoft can publish extensions themselves that do things I can't do as a third party developer, they can use proposed APIs when I can't? Or am I missing something in the proposed API facility? I believe that at least the notebookKernelSource API is still a proposal, it's not that it's been made GA and so those proposed API enablements are not having any effect.

The follow up would be how would I then track that API proposal and work out whether it's likely to be made available in GA and if so when, or what's stopping it being promoted. The proposed API process wiki page doesn't seem to say anything on that front, or at least, not in a way that I understand.

**Comment by gjsjohnmurray** [comment]
An extension that wants to use a particular proposed API must declare it in its package.json. Doing this prevents it being published on Marketplace, so people who want to use it must download the VSIX from somewhere (e.g. an asset of a release created on GitHub). Those users can be using Stable as long as the proposed API is not so new that it is only in Insiders. The users will need to add a command line argument to the launching of VS Code, but a convenient way of achieving this is to add it to their argv.json file. Command Palette has a command to make editing that file easy.

**Reply by tomqwpl** [reply]
Yes, I understand that, that's basically the point of my question. The Microsoft published Jupyter extension is published on the marketplace, and I appear to be using it without providing any additional command line arguments or using the insiders edition, yet it works, and appears to be using proposed APIs. The package.json inside the extension that has been installed in my %USERPROFILE%/.vscode/extensions directory has an "enabledApiProposals" section.

If I clone the source code and run it, it works, I get no complaints about using proposed API. However, if I change the "publisher" field in the package.json to something else, it no longer works, it complains that it is using proposed APIs. So the conclusion seems to be that a) Microsoft allow themselves to publish extensions that use proposed APIs and b) Microsoft published APIs appear to be able to bypass the normal restrictions on having to explicitly permit certain extensions to use proposed APIs on the command line. However so far I've been unable to find the logic in the source code that enables that part. I've also been unable to find anything anywhere that says that Microsoft extensions have this special treatment. Hence the question here. I've also been unable to find anything that says what the process is by which the proposed APIs are promoted (the Wiki page doesn't really give any details there) or how that process is tracked.

**Comment by gjsjohnmurray** [comment]
It's my understanding that certain extensions are "blessed" in a way that means they can use proposed APIs, get published on Marketplace, and not require end-users to take additional steps in order to use them.

**Comment by tomqwpl** [comment, marked as answer]
Ah. I have found the mechanism. In the `product.json` file there is a list under `extensionEnabledApiProposals` of extensions and the list of proposed APIs that they are allowed to use. This is where the list of "blessed" extensions is. It's quite big.

The effect of that though is that those API have to be stable though. In order to change one of them there would need to be a coordinated release of both VSCode and of the extensions using them. If you ever upgraded one and not the other you would then have a broken extension.

**Reply by sameemqiitm** [reply]
Hello,
where is the product.json file ? is the package.json file being referenced here?
If that's the case, even on specifying the  
"enabledApiProposals": [
    "editorInsets"
  ], in package. json , when I share try to run the .vsix file, I still see the error stating that we need to be in extension host development mode  and use the code-insiders . --enable-proposed-api='extension.id'.
  Also this just worked on vs code insiders and not on vs code.
  Is there any way My extension can run on VScode using these proposal api just like Microsoft Extensions (e.g. Copilot,Microsoft Jupyter extension)?

**Reply by eliseevh** [reply]
The product.json file could be found inside your vscode installation (`C:\Users\<user name>\AppData\Local\Programs\Microsoft VS Code\resources\app\product.json`  for me).

**Comment by dk-forgood** [comment]
We have been patiently watching this and related issues through this
year, hoping this API would go public any day. Please, allow us to
adopt it for https://marketplace.visualstudio.com/items?itemName=ZencoderAI.zencoder
It's been well tested in Copilot, so looks pretty stable. We are happy
to move at the dev pace to keep our add-in compatible with the latest
updates.

**Comment by LinZexiao** [comment]
I think the restriction on allowing ordinary developers to use the proposed API should be lifted, because this would enable the proposed API to be tested over a wider range and lead to faster iteration and stabilization. As for the issue that some extensions need to be updated in sync with the proposed API due to the use of the proposed API, this is the price of using an unstable API. The choice should be given to extension developers, rather than prohibiting its use. Although the proposed API can be used to some extent, the threshold has resulted in extensions using the proposed API not being subject to market and user validation at all.
